### PR TITLE
Fix two bugs in async lifting analysis which results in unnecessarily lifted locals.

### LINKED ISSUE #13759: Unnecessarily lifted field in async state machine
**Version Used**: 
2.0.0.60620

**Steps to Reproduce**:
Compile this code:

``` C#
using System.IO;
using System.Threading.Tasks;

class Program
{
    public static void Main() { }

    public static async Task Repro1(Stream source, Stream destination)
    {
        byte[] buffer = new byte[0x1000];
        int bytesRead;
        while ((bytesRead = await source.ReadAsync(buffer, 0, buffer.Length)) != 0)
        {
            await destination.WriteAsync(buffer, 0, bytesRead);
        }
    }

    public static async Task Repro2(Stream source, Stream destination)
    {
        byte[] buffer = new byte[0x1000];
        while (true)
        {
            int bytesRead = await source.ReadAsync(buffer, 0, buffer.Length);
            if (bytesRead == 0) break;
            await destination.WriteAsync(buffer, 0, bytesRead);
        }
    }
}
```

**Expected Behavior**:
The async state machine does not have a field for `bytesRead`.

**Actual Behavior**:
In Repro2, as expected, the async state machine does not lift `bytesRead` to the state machine.  But in Repro1, it does lift `bytesRead` to the state machine, even though the value of `bytesRead` doesn't need to be preserved across an await.

### PR DESCRIPTION
**Pull request by gafter**
The first issue is that we did not properly track the unreachable state in lifting analysis. I believe we did so to work around a bug that has since been fixed.
The second issue is that an assignment in a conditional branch condition was not treated as conditional.

Since VB does not have an assignment expression, I do not have a repro for user locals, but the fix results in some temps no longer being unnecessarily lifted (as shown by changes in existing tests)

### Customer scenario

Async methods have more IL for their bodies, and more locals lifted, than necessary.

### Bugs this fixes

Fixes #13759

### Workarounds, if any

None known.

### Risk

The fix is simple. On the other hand, flow analysis is subtle and the reasons that we tracked unreachable state the way we did is lost in prehistory.

### Performance impact

Changes are simple and local, so no performance impact expected.

### Is this a regression from a previous update?

No.

### Root cause analysis

We do not have extensive testing for which locals are lifted in async code under what circumstances. Too much lifting is generally only a performance (not correctness) problem, so we did not prioritize it.

### How was the bug found?

Customer reported.

### Test documentation updated?

It would be good to add supporting test infrastructure and tests for which locals are lifted for various uses of language constructs. Once that is in place, we could add it as an item to check off for testing new features. This suggestion is recorded in #23929.

### CODE REVIEW

**Comment by gafter** [comment]
test windows_debug_vs-integration_prtest please

**Comment by gafter** [comment]
test windows_debug_vs-integration_prtest please

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews for this fix in lifted variable analysis for async?

File: src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs [hunk 1]
```diff
@@ -2604,9 +2604,19 @@ public override BoundNode VisitArrayLength(BoundArrayLength node)
...
+            if (node.JumpIfTrue)
+            {
+                _pendingBranches.Add(new PendingBranch(node, this.StateWhenTrue));
+                this.SetState(this.StateWhenFalse);
+            }
+            else
+            {
+                _pendingBranches.Add(new PendingBranch(node, this.StateWhenFalse));
+                this.SetState(this.StateWhenTrue);
+            }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs]: Consider choosing state only based on `node.JumpIfTrue`. Same comment for VB.
```
var state = node.JumpIfTrue ? this.StateWhenTrue : this.StateWhenFalse;
_pendingBranches.Add(new PendingBranch(node, state));
this.SetState(state);
```

Comment (gafter) [reply src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs]: No, I think that would not be correct. The jump state and the new current state are supposed to be opposites (one for when the branch is taken, and one for when it is not taken).

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review of this bug fix?

**Comment by gafter** [comment]
Re-targeting to 15.7.

**Pull request merged**